grpc-protocol/src/main/java/com/linecorp/armeria/client/grpc/protocol/UnaryGrpcClient.java

```diff
@@ -239,34 +266,46 @@ public void onNext(DeframedMessage message) {
                         return;
                     }
                     if (processedMessages > 0) {
-                        onError(new ArmeriaStatusException(StatusCodes.INTERNAL,
-                                                           "received more than one data message; " +
-                                                           "UnaryGrpcClient does not support streaming."));
+                        cancel(new ArmeriaStatusException(StatusCodes.INTERNAL,
+                                                          "received more than one data message; " +
+                                                          "UnaryGrpcClient does not support streaming."));
                         return;
                     }
                     final ByteBuf buf = message.buf();
                     // Compression not supported.
                     assert buf != null;
                     content = HttpData.wrap(buf);
+                    buf.retain();
                     processedMessages++;
                 }
 
                 @Override
                 public void onError(Throwable t) {
+                    setCause(t);
                     content.close();
-                    responseFuture.completeExceptionally(t);
+                    responseFuture.completeExceptionally(cause);
```

Comment:
How about making `setCause()` return a `Throwable` so that the relationship between `t` and `cause` are clearer?

```java
final Throwable cause = setCause(t);
content.close();
responseFuture.completeExceptionally(cause);
```

Reply:
I actually refactored how termination is done to align it better with the reactive streams spec, it doesn't store the cause anymore. PTAL.